Fixes #25346 - Correctly render menu for non-admin users

tbrisker opened the pull request:
The new menu implementation in react did not take into account user
permissions as to which menues should be displayed. This also moves the
logic for deciding what to display into the menu classes and removes the
home helper which is no longer used.



<!---

Thank you for contributing to The Foreman, please read the
[following guide](https://www.theforeman.org/contribute.html), in short:

* [Create an issue](https://projects.theforeman.org/projects/foreman/issues)
* Reference the issue via `Fixes #1234` in the commit message
* Prefer present-tense, imperative-style commit messages
* Mark all strings for translation, see [Translating section in the guide]
(https://projects.theforeman.org/projects/foreman/wiki/Translating)
* Prepend `[WIP]` for work in progress to prevent bots from triggering actions
* Be patient, we will do our best to take a look as soon as we can
* Explain the purpose of the PR, attach screenshots if applicable
* Be nice and respectful

We are running bots that will poke you if you miss an item from the list :-)

--->

ohadlevy commented on app/registries/menu/manager.rb:
```diff
@@ -34,7 +34,7 @@ def items(menu_name)
 
       def to_hash(menu_name)
-        items(menu_name).children.map(&:to_hash)
+        items(menu_name).authorized_children.map(&:to_hash)
```
does this work as expected? without reading the code AFAIR this is a registry which might effect all users?

tbrisker replied:
`to_hash` is only called when generating the hash for which menus to display afaict

ohadlevy replied:
would it make sense to create a class instance per user that represent the menus? (in theory might be easy to json (and test) that object vs the menu register)

tbrisker replied:
not sure it's worth the effort, it would take a fairly large refactoring to change it and in any case is outside the scope of this issue (which is a 1.20 blocker). right now the `to_hash`, `authorized?` and `authorized_children` methods are called at runtime with a user context and don't modify the registry and can be tested just fine.

timogoebel reviewed (dismissed):
Works great. Thanks!

timogoebel commented:
@tbrisker: Can you please check out the failing test? Thanks.

tbrisker commented:
@timogoebel thanks, updated the failing test

timogoebel commented:
@tbrisker: Katello failures are related.

HomeHelper is referenced by Katello, see
https://github.com/Katello/katello/blob/b8dabc3d681f9a903d0d53c87c07cf489859083b/app/controllers/katello/application_controller.rb#L6

tbrisker committed:
Fixes #25346 - Correctly render menu for non-admin users

The new menu implementation in react did not take into account user
permissions as to which menues should be displayed. This also moves the
logic for deciding what to display into the menu classes.

tbrisker commented:
hmph... Split off the removal of HomeHelper to GH-6186 and opened https://github.com/Katello/katello/pull/7797 to remove it in katello.

timogoebel commented:
Awesome, APJ!

tbrisker commented:
afaict the katello error is a transient vcr one unrelated to this

ohadlevy commented:
test failures seems not related? (which I would really like to understand how to get rids of it)

tbrisker commented:
indeed, foreman failures are the regular transient ones and katello one is known and should be fixed by now.

timogoebel commented:
Thanks, @tbrisker.

I assume you'll take care of the 🍒-picks.

tbrisker commented:
cherry picked to 1.20 as ee874c2ce
